pombola/south_africa/data/members-interests/pdf_scraper/scrape_interests_pdf.py

```diff
@@ -94,48 +104,164 @@ def scrape_pdf(self):
+                                        self.data[namecount-1][currentsection])-1][curtable[el.attrib['left']]] + text_element
                                 else:
-                                    self.data[namecount-1][currentsection][len(self.data[namecount-1][currentsection])-1][curtable[el.attrib['left']]] = self.data[namecount-1][currentsection][len(self.data[namecount-1][currentsection])-1][curtable[el.attrib['left']]] + ' ' + text_element
+                                    self.data[namecount-1][currentsection][len(self.data[namecount-1][currentsection])-1][curtable[el.attrib['left']]] = self.data[namecount-1][currentsection][len(
+                                        self.data[namecount-1][currentsection])-1][curtable[el.attrib['left']]] + ' ' + text_element
                             else:
-                                self.data[namecount-1][currentsection][len(self.data[namecount-1][currentsection])-1][curtable[el.attrib['left']]] = text_element
+                                self.data[namecount-1][currentsection][len(
+                                    self.data[namecount-1][currentsection])-1][curtable[el.attrib['left']]] = text_element
+
+    def extract_content_from_document(self):
+        """ Extract content from a .docx file and return a (text, html) tuple.
+        """
+        filename = self.input
+        ext = os.path.splitext(filename)[1]
+        if ext == '.docx':
+            with open(filename, "rb") as f:
+                html = mammoth.convert_to_html(f).value
+                # Before returning html, clean it up a bit
+                # ================================
+                # with open(self.output, 'w') as writer:
+                #     writer.writelines(str(html.encode("utf8")))
+                # ================================
+            return html
+        else:
+            # TODO: handle .doc
+            raise ValueError("Can only handle .docx files, but got %s" % ext)
+
+    def parse_html_generated_from_doc(self, html):
+        """
+            Parse the HTML to get the text.
+        """
+        CATEGORIES = {
+            "SHARES AND OTHER FINANCIAL INTERESTS (Family and other trusts)": "SHARES AND OTHER FINANCIAL INTERESTS",
+            "REMUNERATED EMPLOYMENT OUTSIDE PARLIAMENT": "REMUNERATED EMPLOYMENT OUTSIDE PARLIAMENT",
+            "DIRECTORSHIPS AND PARTNERSHIPS": "DIRECTORSHIP AND PARTNERSHIPS",
+            "CONSULTANCIES OR RETAINERSHIPS": "CONSULTANCIES OR RETAINERSHIPS",
+            "SPONSORSHIPS": "SPONSORSHIPS",
+            "GIFTS AND HOSPITALITY": "GIFTS AND HOSPITALITY",
+            "BENEFITS": "BENEFITS",
+            "TRAVEL": "TRAVEL",
+            "LAND AND PROPERTY": "LAND AND PROPERTY",
+            "PENSIONS": "PENSIONS",
+            "PUBLIC CONTRACTS AWARDED": "CONTRACTS",
+            "TRUSTS": "TRUSTS"
+        }
+
+        # state machine
+
+        # state variables
```

Comment:
how did you solve page breaks in the end?

Could you give an example of a page break that you fixed automatically?

It looks like you didn't use the state machine in the end - in that case, let's delete at least the "state machine" comment.

Reply:
Added a comment to explain how the parser is a state machine.